Android Gradle Plugin 7.4.0-beta02 fails with android-cache-fix-gradle-plugin

Environment :
  Gradle : 7.5.1
  AGP : 7.4.0-beta02
  android-cache-fix-gradle-plugin : 2.5.6
  Java : OpenJDK 11

How to reproduce failure
```
git clone https://github.com/ganadist/VersionCodeDemo -b agp74_cache_fix
cd VersionCodeDemo
./gradlew :lib:glide:assemble

> Configure project :lib:glide
WARNING: Android plugin 7.4.0-beta02 has not been tested with this version of the Android cache fix plugin, although it may work.  We test against only the latest patch release versions of Android Gradle plugin: 7.0.4, 7.1.3, 7.2.1, 7.3.0-beta05, 7.4.0-alpha08.  If 7.4.0-beta02 is newly released, we may not have had a chance to release a version tested against it yet.  Proceed with caution.  You can suppress this warning with with -Dorg.gradle.android.cache-fix.ignoreVersionCheck=true.
Values of variant API AnnotationProcessorOptions.arguments are queried and may return non final values, this is unsupported

FAILURE: Build failed with an exception.

* What went wrong:
A problem occurred configuring project ':lib:glide'.
> com.android.build.gradle.internal.crash.ExternalApiUsageException: com.android.build.gradle.internal.dsl.AgpDslLockedException: It is too late to modify AnnotationProcessor::argumentProviders
  It has already been read to configure this project.
  Consider either moving this call to be during evaluation,
  or using the variant API.


Caused by: com.android.build.gradle.internal.dsl.AgpDslLockedException: It is too late to modify AnnotationProcessor::argumentProviders	
It has already been read to configure this project.	
Consider either moving this call to be during evaluation,	
or using the variant API.	
	at com.android.build.gradle.internal.dsl.decorator.LockableList.add(LockableList.kt:83)	
		•••
	at org.gradle.android.workarounds.RoomSchemaLocationWorkaround$_apply_closure3$_closure14.doCall(RoomSchemaLocationWorkaround.groovy:137)	
		•••
	at com.android.build.gradle.LibraryExtension.addVariant(LibraryExtension.kt:102)	
	at com.android.build.gradle.internal.ApiObjectFactory.create(ApiObjectFactory.java:119)	
	at com.android.build.gradle.internal.plugins.BasePlugin.createAndroidTasks(BasePlugin.kt:706)	
	at com.android.build.gradle.internal.plugins.BasePlugin$createTasks$2$1.call(BasePlugin.kt:572)	
	at com.android.build.gradle.internal.profile.NoOpAnalyticsConfiguratorService.recordBlock(NoOpAnalyticsConfiguratorService.kt:53)	
	at com.android.build.gradle.internal.plugins.BasePlugin$createTasks$2.accept(BasePlugin.kt:567)	
	at com.android.build.gradle.internal.plugins.BasePlugin$createTasks$2.accept(BasePlugin.kt:564)	
	at com.android.build.gradle.internal.crash.CrashReporting$afterEvaluate$1.execute(crash_reporting.kt:37)	
	at com.android.build.gradle.internal.crash.CrashReporting$afterEvaluate$1.execute(crash_reporting.kt:35)
```

--- a/src/main/groovy/org/gradle/android/workarounds/RoomSchemaLocationWorkaround.groovy
+++ b/src/main/groovy/org/gradle/android/workarounds/RoomSchemaLocationWorkaround.groovy
@@ -91,7 +91,7 @@ class RoomSchemaLocationWorkaround implements Workaround {
                     // Make sure that the annotation processor argument has not been explicitly configured in the Android
                     // configuration (i.e. we only want this configured through the room extension)
                     Map<String, String> arguments = variant.javaCompileOptions.annotationProcessorOptions.arguments
-                    errorIfAnnotationProcessorSet(arguments.keySet())
+                    errorIfRoomSchemaAnnotationArgumentSet(arguments.keySet())
 
                     def variantSpecificSchemaDir = project.objects.directoryProperty()
                     variantSpecificSchemaDir.set(getVariantSpecificSchemaDir(project, "compile${variant.name.capitalize()}JavaWithJavac"))
@@ -111,7 +111,7 @@ class RoomSchemaLocationWorkaround implements Workaround {
                     // Make sure that the annotation processor argument has not been explicitly configured in the Android
                     // configuration (i.e. we only want this configured through the room extension
                     MapProperty<String, String> arguments = variant.javaCompilation.annotationProcessor.arguments
-                    errorIfAnnotationProcessorSet(arguments.keySet().get())
+                    errorIfRoomSchemaAnnotationArgumentSet(arguments.keySet().get())
 
                     def variantSpecificSchemaDir = project.objects.directoryProperty()
                     variantSpecificSchemaDir.set(getVariantSpecificSchemaDir(project, "compile${variant.name.capitalize()}JavaWithJavac"))
@@ -225,7 +225,7 @@ class RoomSchemaLocationWorkaround implements Workaround {
         }
     }
 
-    private static void errorIfAnnotationProcessorSet(Set<String> options) {
+    private static void errorIfRoomSchemaAnnotationArgumentSet(Set<String> options) {
         if (options.contains(ROOM_SCHEMA_LOCATION)) {
             throw new IllegalStateException("""${RoomSchemaLocationWorkaround.class.name} cannot be used with an explicit '${ROOM_SCHEMA_LOCATION}' annotation processor argument.  Please change this to configure the schema location directory via the 'room' project extension:
     room {

--- a/src/main/groovy/org/gradle/android/workarounds/RoomSchemaLocationWorkaround.groovy
+++ b/src/main/groovy/org/gradle/android/workarounds/RoomSchemaLocationWorkaround.groovy
@@ -2,6 +2,7 @@ package org.gradle.android.workarounds
 
 import org.gradle.android.AndroidIssue
 import org.gradle.android.VersionNumber
+import org.gradle.android.Versions
 import org.gradle.api.DefaultTask
 import org.gradle.api.Project
 import org.gradle.api.Task
@@ -12,6 +13,7 @@ import org.gradle.api.file.DirectoryProperty
 import org.gradle.api.file.DuplicatesStrategy
 import org.gradle.api.internal.file.FileOperations
 import org.gradle.api.model.ObjectFactory
+import org.gradle.api.provider.MapProperty
 import org.gradle.api.provider.Provider
 import org.gradle.api.tasks.Internal
 import org.gradle.api.tasks.Optional
@@ -24,7 +26,6 @@ import org.gradle.work.DisableCachingByDefault
 
 import javax.inject.Inject
 import java.lang.reflect.Field
-import java.util.function.Supplier
 
 /**
  * Changes annotation processor arguments to set the room.schemaLocation via a CommandLineArgsProcessor
@@ -74,9 +75,6 @@ class RoomSchemaLocationWorkaround implements Workaround {
         // Grab fileOperations so we can do copy/sync operations
         def fileOperations = project.fileOperations
 
-        // An undefined directory property for use when providers are disabled
-        def nullDirectory = project.objects.directoryProperty()
-
         // Create a task that will be used to merge the task-specific schema locations to the directory (or directories)
         // originally specified.  This allows us to fan out the generated output and keep good cacheability for the
         // compile/kapt tasks but still join everything later in the location the user expects.
@@ -86,57 +84,105 @@ class RoomSchemaLocationWorkaround implements Workaround {
 
         boolean javaCompileSchemaGenerationEnabled = true
 
-        def configureVariant = { variant ->
-            // Make sure that the annotation processor argument has not been explicitly configured in the Android
-            // configuration (i.e. we only want this configured through the room extension
-            Map<String, String> arguments = variant.javaCompileOptions.annotationProcessorOptions.arguments
-            if (arguments.containsKey(ROOM_SCHEMA_LOCATION)) {
-                throw new IllegalStateException("""${this.class.name} cannot be used with an explicit '${ROOM_SCHEMA_LOCATION}' annotation processor argument.  Please change this to configure the schema location directory via the 'room' project extension:
-    room {
-        schemaLocationDir.set(file("roomSchemas"))
-    }
-""")
+        applyToAllAndroidVariants(project, new ConfigureVariants() {
+            @Override
+            Closure<?> getOldVariantConfiguration() {
+                return { variant ->
+                    // Make sure that the annotation processor argument has not been explicitly configured in the Android
+                    // configuration (i.e. we only want this configured through the room extension)
+                    Map<String, String> arguments = variant.javaCompileOptions.annotationProcessorOptions.arguments
+                    errorIfAnnotationProcessorSet(arguments.keySet())
+
+                    def variantSpecificSchemaDir = project.objects.directoryProperty()
+                    variantSpecificSchemaDir.set(getVariantSpecificSchemaDir(project, "compile${variant.name.capitalize()}JavaWithJavac"))
+
+                    // Add a command line argument provider to the compile task argument providers
+                    variant.javaCompileProvider.configure { JavaCompile task ->
+                        task.options.compilerArgumentProviders.add(
+                            new JavaCompilerRoomSchemaLocationArgumentProvider(roomExtension.schemaLocationDir, variantSpecificSchemaDir)
+                        )
+                    }
+                }
             }
 
-            // Configure the annotation processor argument provider on the Java compile task
-            javaCompileProvider.configure { JavaCompile task ->
-                def taskSpecificSchemaDir = project.objects.directoryProperty()
-                taskSpecificSchemaDir.set(getTaskSpecificSchemaDir(task))
+            @Override
+            Closure<?> getNewVariantConfiguration() {
+                return { variant ->
+                    // Make sure that the annotation processor argument has not been explicitly configured in the Android
+                    // configuration (i.e. we only want this configured through the room extension
+                    MapProperty<String, String> arguments = variant.javaCompilation.annotationProcessor.arguments
+                    errorIfAnnotationProcessorSet(arguments.keySet().get())
 
-                // Add a command line argument provider to the task-specific list of providers
-                task.options.compilerArgumentProviders.add(
-                    new JavaCompilerRoomSchemaLocationArgumentProvider(roomExtension.schemaLocationDir, taskSpecificSchemaDir, nullDirectory, { javaCompileSchemaGenerationEnabled } as Supplier)
-                )
+                    def variantSpecificSchemaDir = project.objects.directoryProperty()
+                    variantSpecificSchemaDir.set(getVariantSpecificSchemaDir(project, "compile${variant.name.capitalize()}JavaWithJavac"))
 
-                // Register the generated schemas to be merged back to the original specified schema directory
-                task.project.gradle.taskGraph.whenReady { TaskExecutionGraph graph ->
-                    if (graph.hasTask(task)) {
-                        roomExtension.registerOutputDirectory(taskSpecificSchemaDir)
-                    }
+                    // Add a command line argument provider to the variant list of providers
+                    variant.javaCompilation.annotationProcessor.argumentProviders.add(
+                        new JavaCompilerRoomSchemaLocationArgumentProvider(roomExtension.schemaLocationDir, variantSpecificSchemaDir)
+                    )
                 }
+            }
+        })
 
-                // Seed the task-specific generated schema dir with the existing schemas
-                task.doFirst {
+        project.tasks.withType(JavaCompile).configureEach { task ->
+            TaskExecutionGraph taskGraph = task.project.gradle.taskGraph
+
+            taskGraph.whenReady {
+                if (taskGraph.hasTask(task)) {
                     if (javaCompileSchemaGenerationEnabled) {
-                        RoomSchemaLocationWorkaround.copyExistingSchemasToTaskSpecificTmpDir(fileOperations, roomExtension.schemaLocationDir, taskSpecificSchemaDir)
+                        // Seed the task-specific generated schema dir with the existing schemas
+                        task.doFirst onlyIfAnnotationProcessorConfiguredForJavaCompile(task.options.compilerArgumentProviders) { JavaCompilerRoomSchemaLocationArgumentProvider provider ->
+                            RoomSchemaLocationWorkaround.copyExistingSchemasToTaskSpecificDirForJavaCompile(fileOperations, roomExtension.schemaLocationDir, provider)
+                        }
+
+                        // Register the generated schemas to be merged back to the original specified schema directory
+                        task.configure onlyIfAnnotationProcessorConfiguredForJavaCompile(task.options.compilerArgumentProviders) { JavaCompilerRoomSchemaLocationArgumentProvider provider ->
+                            roomExtension.registerOutputDirectory(provider.schemaLocationDir)
+                        }
+                    } else {
+                        // If kapt is enabled, then those tasks will do the annotation processing, and we should go through
+                        // and remove the provider from the JavaCompile tasks since we don't want their outputs to be considered
+                        def itr = task.options.compilerArgumentProviders.iterator()
+                        while (itr.hasNext()) {
+                            CommandLineArgumentProvider provider = itr.next()
+                            if (provider instanceof RoomSchemaLocationArgumentProvider) {
+                                itr.remove()
+                            }
+                        }
                     }
                 }
+            }
 
-                task.finalizedBy { roomExtension.schemaLocationDir.isPresent() ? mergeTask : null }
+            task.finalizedBy {
+                onlyIfAnnotationProcessorConfiguredForJavaCompile(task.options.compilerArgumentProviders) {
+                    roomExtension.schemaLocationDir.isPresent() ? mergeTask : null
+                }
             }
         }
-
-        applyToAllAndroidVariants(project, configureVariant)
 
         project.plugins.withId("kotlin-kapt") {
             // The kapt task has a list of annotation processor providers which _is_ the list of providers
             // in the Android variant, so we can't just add a task-specific provider.  To handle kapt tasks,
             // we _have_ to add the task-specific provider to the variant.
-            applyToAllAndroidVariants(project) { variant ->
-                def variantSpecificSchemaDir = project.objects.directoryProperty()
-                variantSpecificSchemaDir.set(getVariantSpecificSchemaDir(project, "kapt${variant.name.capitalize()}Kotlin"))
-                variant.javaCompileOptions.annotationProcessorOptions.compilerArgumentProviders.add(new KaptRoomSchemaLocationArgumentProvider(roomExtension.schemaLocationDir, variantSpecificSchemaDir, nullDirectory))
-            }
+            applyToAllAndroidVariants(project, new ConfigureVariants() {
+                @Override
+                Closure<?> getOldVariantConfiguration() {
+                    return { variant ->
+                        def variantSpecificSchemaDir = project.objects.directoryProperty()
+                        variantSpecificSchemaDir.set(getVariantSpecificSchemaDir(project, "kapt${variant.name.capitalize()}Kotlin"))
+                        variant.javaCompileOptions.annotationProcessorOptions.compilerArgumentProviders.add(new KaptRoomSchemaLocationArgumentProvider(roomExtension.schemaLocationDir, variantSpecificSchemaDir))
+                    }
+                }
+
+                @Override
+                Closure<?> getNewVariantConfiguration() {
+                    return { variant ->
+                        def variantSpecificSchemaDir = project.objects.directoryProperty()
+                        variantSpecificSchemaDir.set(getVariantSpecificSchemaDir(project, "kapt${variant.name.capitalize()}Kotlin"))
+                        variant.javaCompilation.annotationProcessor.argumentProviders.add(new KaptRoomSchemaLocationArgumentProvider(roomExtension.schemaLocationDir, variantSpecificSchemaDir))
+                    }
+                }
+            })
 
             // Kapt tasks will remove the contents of any output directories, which will interfere with any additional
             // annotation processors that use the room schema location processor argument and expect existing schemas to
@@ -179,6 +225,24 @@ class RoomSchemaLocationWorkaround implements Workaround {
         }
     }
 
+    private static void errorIfAnnotationProcessorSet(Set<String> options) {
+        if (options.contains(ROOM_SCHEMA_LOCATION)) {
+            throw new IllegalStateException("""${RoomSchemaLocationWorkaround.class.name} cannot be used with an explicit '${ROOM_SCHEMA_LOCATION}' annotation processor argument.  Please change this to configure the schema location directory via the 'room' project extension:
+    room {
+        schemaLocationDir.set(file("roomSchemas"))
+    }
+""")
+        }
+    }
+    private static Closure onlyIfAnnotationProcessorConfiguredForJavaCompile(def argumentProviders, Closure<?> action) {
+        return {
+            def provider = argumentProviders.find { it instanceof JavaCompilerRoomSchemaLocationArgumentProvider }
+            if (provider != null) {
+                action.call(provider)
+            }
+        }
+    }
+
     private static Closure onlyIfAnnotationProcessorConfiguredForKapt(def annotationProcessorOptionProviders, Closure<?> action) {
         return {
             def provider = annotationProcessorOptionProviders.flatten().find { it instanceof KaptRoomSchemaLocationArgumentProvider }
@@ -188,23 +252,32 @@ class RoomSchemaLocationWorkaround implements Workaround {
         }
     }
 
-    private static void applyToAllAndroidVariants(Project project, Closure<?> configureVariant) {
+    private static void applyToAllAndroidVariants(Project project, ConfigureVariants configureVariants) {
+        if (Versions.CURRENT_ANDROID_VERSION <= VersionNumber.parse("7.4.0-alpha01")) {
+            applyToAllAndroidVariantsWithOldVariantApi(project, configureVariants)
+        } else {
+            applyToAllAndroidVariantsWithNewVariantApi(project, configureVariants)
+        }
+    }
+
+    private static void applyToAllAndroidVariantsWithOldVariantApi(Project project, ConfigureVariants configureVariants) {
         project.plugins.withId("com.android.application") {
             def android = project.extensions.findByName("android")
 
-            android.applicationVariants.all(configureVariant)
+            android.applicationVariants.all(configureVariants.oldVariantConfiguration)
         }
 
         project.plugins.withId("com.android.library") {
             def android = project.extensions.findByName("android")
 
-            android.libraryVariants.all(configureVariant)
+            android.libraryVariants.all(configureVariants.oldVariantConfiguration)
         }
     }
 
-    static File getTaskSpecificSchemaDir(Task task) {
-        def schemaBaseDir = task.project.layout.buildDirectory.dir("roomSchemas").get().asFile
-        return new File(schemaBaseDir, task.name)
+    private static void applyToAllAndroidVariantsWithNewVariantApi(Project project, ConfigureVariants configureVariants) {
+        def androidComponents = project.extensions.findByName("androidComponents")
+        def selector = androidComponents.selector()
+        androidComponents.onVariants(selector.all(), configureVariants.newVariantConfiguration)
     }
 
     static File getVariantSpecificSchemaDir(Project project, String variantName) {
@@ -223,6 +296,14 @@ class RoomSchemaLocationWorkaround implements Workaround {
                 it.into(taskSpecificTmpDir)
             }
         }
+    }
+
+    private static void copyExistingSchemasToTaskSpecificDirForJavaCompile(FileOperations fileOperations, Provider<Directory> existingSchemaDir, JavaCompilerRoomSchemaLocationArgumentProvider provider) {
+        // Derive the variant directory from the command line provider it is configured with
+        def variantSpecificSchemaDir = provider.schemaLocationDir
+
+        // Populate the variant-specific temporary schema dir with the existing schemas
+        copyExistingSchemasToTaskSpecificTmpDir(fileOperations, existingSchemaDir, variantSpecificSchemaDir)
     }
 
     private static void copyExistingSchemasToTaskSpecificTmpDirForKapt(FileOperations fileOperations, Provider<Directory> existingSchemaDir, KaptRoomSchemaLocationArgumentProvider provider) {
@@ -305,17 +386,9 @@ class RoomSchemaLocationWorkaround implements Workaround {
         @Internal
         final Provider<Directory> schemaLocationDir
 
-        @Internal
-        final Supplier<Boolean> enabled
-
-        @Internal
-        final Provider<Directory> nullDirectory
-
-        RoomSchemaLocationArgumentProvider(Provider<Directory> configuredSchemaLocationDir, Provider<Directory> schemaLocationDir, Provider<Directory> nullDirectory, Supplier<Boolean> enabled) {
+        RoomSchemaLocationArgumentProvider(Provider<Directory> configuredSchemaLocationDir, Provider<Directory> schemaLocationDir) {
             this.configuredSchemaLocationDir = configuredSchemaLocationDir
-            this.enabled = enabled
             this.schemaLocationDir = schemaLocationDir
-            this.nullDirectory = nullDirectory
         }
 
         @Internal
@@ -325,7 +398,7 @@ class RoomSchemaLocationWorkaround implements Workaround {
 
         @Override
         Iterable<String> asArguments() {
-            if (configuredSchemaLocationDir.isPresent() && enabled.get()) {
+            if (configuredSchemaLocationDir.isPresent()) {
                 return ["-A${ROOM_SCHEMA_LOCATION}=${schemaLocationPath}" as String]
             } else {
                 return []
@@ -335,21 +408,21 @@ class RoomSchemaLocationWorkaround implements Workaround {
         @OutputDirectory
         @Optional
         Provider<Directory> getEffectiveSchemaLocationDir() {
-            return enabled.get() ? schemaLocationDir : nullDirectory
+            return schemaLocationDir
         }
     }
 
     static class JavaCompilerRoomSchemaLocationArgumentProvider extends RoomSchemaLocationArgumentProvider {
-        JavaCompilerRoomSchemaLocationArgumentProvider(Provider<Directory> configuredSchemaLocationDir, Provider<Directory> schemaLocationDir, Provider<Directory> nullDirectory, Supplier<Boolean> enabled) {
-            super(configuredSchemaLocationDir, schemaLocationDir, nullDirectory, enabled)
+        JavaCompilerRoomSchemaLocationArgumentProvider(Provider<Directory> configuredSchemaLocationDir, Provider<Directory> schemaLocationDir) {
+            super(configuredSchemaLocationDir, schemaLocationDir)
         }
     }
 
     static class KaptRoomSchemaLocationArgumentProvider extends RoomSchemaLocationArgumentProvider {
         private Provider<Directory> temporarySchemaLocationDir
 
-        KaptRoomSchemaLocationArgumentProvider(Provider<Directory> configuredSchemaLocationDir, Provider<Directory> schemaLocationDir, Provider<Directory> nullDirectory) {
-            super(configuredSchemaLocationDir, schemaLocationDir, nullDirectory, { true } as Supplier<Boolean>)
+        KaptRoomSchemaLocationArgumentProvider(Provider<Directory> configuredSchemaLocationDir, Provider<Directory> schemaLocationDir) {
+            super(configuredSchemaLocationDir, schemaLocationDir)
             this.temporarySchemaLocationDir = schemaLocationDir.map {it.dir("../${it.asFile.name}Temp") }
         }
 
@@ -405,5 +478,21 @@ class RoomSchemaLocationWorkaround implements Workaround {
                 }
             }
         }
+    }
+
+    /**
+     * Represents equivalent configurations for AGP versions that support either the old or new variant APIs.
+     */
+    private interface ConfigureVariants {
+        /**
+         * The equivalent configuration for the old variant API.  Closure will receive a {@link com.android.build.gradle.api.BaseVariant}
+         * as the argument.
+         */
+        Closure<?> getOldVariantConfiguration();
+        /**
+         * The equivalient configuration for the new variant API.  Closure will receive a {@link com.android.build.api.variant.Variant}
+         * as the argument.
+         */
+        Closure<?> getNewVariantConfiguration();
     }
 }

--- a/src/test/groovy/org/gradle/android/RoomSchemaLocationWorkaroundTest.groovy
+++ b/src/test/groovy/org/gradle/android/RoomSchemaLocationWorkaroundTest.groovy
@@ -4,6 +4,7 @@ import org.gradle.android.workarounds.RoomSchemaLocationWorkaround
 import org.gradle.testkit.runner.BuildResult
 import org.gradle.testkit.runner.TaskOutcome
 import org.junit.Assume
+import spock.lang.Issue
 import spock.lang.Unroll
 
 import static org.gradle.testkit.runner.TaskOutcome.FROM_CACHE
@@ -455,6 +456,59 @@ class RoomSchemaLocationWorkaroundTest extends AbstractTest {
         where:
         //noinspection GroovyAssignabilityCheck
         androidVersion << TestVersions.latestAndroidVersions
+    }
+
+    @Issue("https://github.com/gradle/android-cache-fix-gradle-plugin/issues/353")
+    @Unroll
+    def "does not error when tasks are eagerly created (Android #androidVersion) (Kotlin #kotlinVersion)"() {
+        def kotlinVersionNumber = VersionNumber.parse(kotlinVersion)
+        // There are kotlin module version errors when using older versions of kotlin with AGP 7.2.0+ in this configuration
+        Assume.assumeFalse(androidVersion >= VersionNumber.parse("7.2.0-alpha01") && kotlinVersionNumber < VersionNumber.parse("1.5.0"))
+
+        SimpleAndroidApp.builder(temporaryFolder.root, cacheDir)
+            .withAndroidVersion(androidVersion)
+            .withKotlinVersion(VersionNumber.parse(kotlinVersion))
+            .build()
+            .writeProject()
+
+        file('app/build.gradle') << eagerlyCreateJavaCompileTasks
+        file('library/build.gradle') << eagerlyCreateJavaCompileTasks
+
+        cacheDir.deleteDir()
+        cacheDir.mkdirs()
+
+        when:
+        BuildResult buildResult = withGradleVersion(TestVersions.latestSupportedGradleVersionFor(androidVersion).version)
+            .forwardOutput()
+            .withProjectDir(temporaryFolder.root)
+            .withArguments(CLEAN_BUILD)
+            .build()
+
+        then:
+        assertCompileTasksHaveOutcome(buildResult, SUCCESS)
+        assertCompileAndroidTestTasksHaveOutcome(buildResult, SUCCESS)
+        assertCompileUnitTestTasksHaveOutcome(buildResult, SUCCESS)
+        assertKaptTasksHaveOutcome(buildResult, SUCCESS)
+        assertKaptAndroidTestTasksHaveOutcome(buildResult, SUCCESS)
+        assertKaptUnitTestTasksHaveOutcome(buildResult, SUCCESS)
+        buildResult.task(':app:mergeRoomSchemaLocations').outcome == SUCCESS
+        buildResult.task(':library:mergeRoomSchemaLocations').outcome == SUCCESS
+
+        and:
+        assertKaptSchemaOutputsExist()
+
+        and:
+        assertMergedSchemaOutputsExist()
+
+        where:
+        //noinspection GroovyAssignabilityCheck
+        [androidVersion, kotlinVersion] << [TestVersions.latestAndroidVersions, TestVersions.supportedKotlinVersions].combinations()
+    }
+
+    private static String getEagerlyCreateJavaCompileTasks() {
+        """
+            tasks.withType(JavaCompile) { println it.path }
+        """
     }
 
     void assertNotExecuted(buildResult, String taskPath) {

